## <span style=color:brown;> User-Defined Functions in Python </span>

* In Python, a **function** is a block of reusable code that is used to perform a specific task.  
* You can create your own custom functions, known as user-defined functions, to organize your code, make it more readable, and avoid repetition.  
* Functions are a core part of programming as they allow for modular, reusable, and organized code.

**What is a Function?**

A function is a block of organized, reusable code that performs a single action or task. Functions provide better modularity and reusability to your code.  

There are two types of functions in Python:

Built-in Functions: Functions like print(), len(), etc.  
User-Defined Functions: Functions that you, the programmer, define to meet your needs.

**Why Use Functions?**

Functions allow you to:

* Organize your code: Break your code into smaller, manageable parts.
* Reusability: Write a piece of code once and use it multiple times.
* Maintainability: Make your code easier to maintain and update.
* Avoid repetition: Reduce redundancy by avoiding the repetition of code

Creating a Function (Syntax)  

In Python, you create a function using the **def** keyword, followed by the function name and parentheses. The code inside the function is indented.

In [ ]:
def function_name(parameters):
    # Code block (function body)
    return value  # Optional

* def: Keyword to define a function.
* function_name: The name you give to the function.
* parameters: (Optional) Values passed to the function when it is called.
* return: (Optional) Used to return a value from the function.

##### Defining a Simple Function

In [1]:
def greet():
    print("Hello, welcome to Python!")

In [2]:
greet()

Hello, welcome to Python!


**Calling a Function**

Once a function is defined, you can call it by using its name followed by parentheses.

In [4]:
def greet():
    print("Hello, welcome to Python!")

# Call the function
greet()

Hello, welcome to Python!


**Function Parameters and Arguments**

You can pass values to a function when you call it. These values are known as arguments and are received by the function as parameters.

**Function with Parameters**

In [5]:
def greet_user(name):
    print(f"Hello, {name}!")

# Call the function with an argument
greet_user("Alice")

Hello, Alice!


* name is the parameter.
* "Alice" is the argument passed to the function.

**Return Values**
  
A function can return a value back to the caller using the return statement. Once return is called, the function exits.

In [15]:
def add_numbers(a, b):
    result = a + b
    return result

In [16]:
add_numbers(10,5)

15

* The function add_numbers() returns the sum of a and b.

**No Parameters, No Return Value**

In [7]:
def say_hello():
    print("Hello, world!")

say_hello()

Hello, world!


**Parameters, No Return Value**

In [8]:
def greet(name):
    print(f"Hi {name}, welcome!")

greet("Bob")

Hi Bob, welcome!


**Parameters and Return Value**

In [9]:
def multiply(a, b):
    return a * b

product = multiply(2, 3)
print("Product:", product)

Product: 6


**Keyword and Default Arguments**

You can pass arguments by their keyword. This allows you to assign values based on the parameter names, regardless of the order.

In [10]:
def display_info(name, age):
    print(f"Name: {name}, Age: {age}")

In [11]:
# Calling the function with keyword arguments
display_info(age=25, name="John")

Name: John, Age: 25


### Practical Examples of User-Defined Functions

**Function to Check if a Number is Even or Odd**

In [17]:
def is_even(number):
    if number % 2 == 0:
        return True
    else:    
        return False

In [18]:
is_even(4)

True

In [19]:
is_even(7)

False

**Function to Calculate the Area of a Circle**

In [23]:
def circle_areaa(radius):
    return 3.14 * radius ** 2

In [24]:
circle_areaa(5)

78.5

### Recursion

Recursion is a programming technique where a function calls itself to solve a smaller instance of the same problem. This continues until it reaches a base case, which is a condition that stops the recursion.

Simply put, recursion breaks a problem down into smaller sub-problems. Each recursive call works on a smaller part of the problem until a solution is found.

**Factorial Using Recursion**

The factorial of a number (n!) is the product of all positive integers less than or equal to that number. For example, 5! = 5 * 4 * 3 * 2 * 1.

In [20]:
def factorial(n):
    # Base case: when n is 1, stop recursion
    if n == 1:
        return 1
    # Recursive case: multiply n by the factorial of (n-1)
    else:
        return n * factorial(n - 1)

In [21]:
factorial(5)

120

**Explanation:**

If n is 1, the function returns 1 (base case).  
Otherwise, the function calls itself with n-1 until it reaches the base case.  
For factorial(5), it computes:  
5 * factorial(4)  
4 * factorial(3)  
3 * factorial(2)  
2 * factorial(1)  
Finally, it returns 120.

**Calculating the sum of numbers from 1 to n**

We want to find the sum of all numbers from 1 to a given number n. For example, if n = 5, the sum is 1 + 2 + 3 + 4 + 5 = 15

We can solve this using recursion by breaking it into smaller problems:

The sum of numbers from 1 to n is equal to n + sum(1 to n-1).

In [1]:
def recursive_sum(n):
    # Base case: if n is 1, return 1 (this stops the recursion)
    if n == 1:
        return 1
    # Recursive case: return n plus the sum of numbers from 1 to n-1
    else:
        return n + recursive_sum(n - 1)

In [2]:
recursive_sum(5)

15

**Explanation:**

Base case: If n is 1, we simply return 1. This is the simplest form of the problem.  

Recursive case: If n is greater than 1, we return n + recursive_sum(n - 1). This keeps calling the function with smaller numbers until it reaches the base case.`

### Summary
  
* Functions in Python are defined using the def keyword.
* Functions allow for code reusability and modularity.
* Arguments and parameters

***